src/components/implementation/fork_tests/voter/voter.c

```diff
@@ -0,0 +1,290 @@
+#include <cos_component.h>
+#include <print.h>
+#include <quarantine.h>
+#include <cbuf.h>
+#include <sched.h>
+#include <voter.h>
+#include <periodic_wake.h>
+
+struct replica_info {
+	spdid_t spdid;
+	cbuf_t read_buffer;
+	cbuf_t write_buffer;
+	void *buf_read;
+	void *buf_write;
+	unsigned short int thread_id;
+	int blocked;
+};
+
+struct nmod_comp {
+	struct replica_info *replicas;
+	int nreplicas;
+};
+
+struct channel {
+	struct nmod_comp *A, *B;	// come up with better terms, they're symmetric
+	int timer; // is this an int, figure it out later
+	int A2B_data, have_A2B_data;	// make these arrays
+	int B2A_data, have_B2A_data;
+};
+
+/* number of components and component array */
+int n_comps = 2;
+struct nmod_comp components[2];
+
+/* number of channels and channels array. Symmetric. */
+int n_channels = 1;
+struct channel channels[1];
+
+int period = 100;
+
+/* map of spd to replica type */
+replica_type map[250];
+
+void cos_fix_spdid_metadata(spdid_t o_spd, spdid_t f_spd) { }
+
+struct nmod_comp *get_component(spdid_t spdid) {
+	replica_type type = map[spdid];
+	int i;
+
+	if (type == none) return NULL;
+
+	// there is definitely a less stupid way to do this...
+	if (type == ping) {
+		for (i = 0; i < components[0].nreplicas; i++) {
+			if (components[0].replicas[i].spdid == spdid) return &components[0];
+		}
+		return NULL;
+	}
+	else if (type == pong) {
+		for (i = 0; i < components[1].nreplicas; i++) {
+			if (components[1].replicas[i].spdid == spdid) return &components[1];
+		}
+		return NULL;
+	}
+	else {
+		return NULL;
+	}
+}
+
+struct replica_info *get_replica(spdid_t spdid) {
+	struct nmod_comp *component = NULL;
+	int i;
+	replica_type type = map[spdid];
+	if (type == none) return NULL;
+
+	component = get_component(spdid);
+	if (!component) return NULL;
+
+	for (i = 0; i < component->nreplicas; i++) {
+		if (component->replicas[i].spdid == spdid) return &component->replicas[i];
```

Comment:
again, replace all of this with a map from spdid to replica (and perhaps have poi8nters in the replicas to the components etc...)